File: generate_document/generateDocument.go
package main

import "fmt"

func generateDocument(chars string, doc string) bool {
	if len(doc) > 0 {
		qty := make(map[rune]int)

		for _, ch := range chars {
			if _, ok := qty[ch]; ok {
				qty[ch] += 1
			} else {
				qty[ch] = 1
			}
		}
		for _, ch := range doc {
			if v, ok := qty[ch]; ok {
				if v-1 >= 0 {
					qty[ch] -= 1
				} else {
					return false
				}
			} else {
				return false
			}
		}
	}
	return true
}

func main() {
	characters := "Bste!hetsi ogEAxpelrt x "
	document := "AlgoExpert is the Best!"
	fmt.Println(true, "vs", generateDocument(characters, document))
}

File: generate_document/generate_document.py
#Complexity: O(n + m) n - length of document, m - number of characters
#Space: O(c) c - number of unique characters
def generateDocument(characters, document):
    if len(document):
        buf = dict()
        for c in characters:
            buf[c] = buf.get(c, 0) + 1
        for c in document:
            if c not in buf or buf[c] - 1 < 0:
                return False
            buf[c] -= 1
    return True


if __name__ == "__main__":
    characters = "Bste!hetsi ogEAxpelrt x "
    document = "AlgoExpert is the Best!"
    print(True, "vs", generateDocument(characters, document))
